src/filter.jl

```diff
@@ -85,3 +85,206 @@ function hamilton_filter(y::AbstractVector, h::Integer)
+- `p`: Likelihood of data at period `t`.
+"""
+function update!(p_s_update::AbstractVector, rsm::RegimeSwitchingModel, p_s_forecast::AbstractArray, t::Integer)
+    y_at_period_t = get_y_at_period_t(rsm.y, t)
+    eta = [max(0, rsm.g(y_at_period_t, s, rsm.parameter)) for s in 1:rsm.M]
+    tmp = eta .* p_s_forecast
+    p = sum(tmp)
+    p_s_update .= tmp./p
+    return p
+end
+get_y_at_period_t(y::AbstractMatrix, t::Union{Integer, AbstractVector}) = y[t, :]
+get_y_at_period_t(y::AbstractVector, t::Union{Integer, AbstractVector}) = y[t]
+
+"""
+Apply the backward smoother developed by Kim to a regime switching model.
+
+##### Arguments
+- `rsm::RegimeSwitchingModel`: `RegimeSwitchingModel` specifying the model.
+- `prob_update_pre::AbstractArray`: Probability distribution of state at period 0 conditional on the information up to
+                                   period 0.
+- `verbose::Bool`: If `true`, the funciton returns joint distribution. If `false`, `nothing` is returned.
+
+##### Returns (only when `verbose` is `true`)
+- `p_s_joint_smoothed`: Joint distribution of state at period `t` and `t+1` data conditional on the all information.
+"""
+function smooth!(rsm::RegimeSwitchingModel;
+                 prob_update_pre::AbstractArray = stationary_distributions(MarkovChain(rsm.P))[1],
+                 verbose::Bool=false)
+    p_s_joint_smoothed = Array{Float64, 3}(undef, size(rsm.y, 1), rsm.M, rsm.M)
+    smooth!(rsm, p_s_joint_smoothed, prob_update_pre = prob_update_pre)
+    verbose ? (return p_s_joint_smoothed) : (return nothing) 
+end
+
+"""
+Apply the backward smoother developed by Kim to a regime switching model.
+
+Same as `smooth` except that the result is stored in the perallocated first argument.
+
+##### Arguments
+- `rsm::RegimeSwitchingModel`: `RegimeSwitchingModel` specifying the model.
+- `p_s_joint_smoothed`: `T x n_state x n_state`
+- `prob_update_pre::AbstractArray`: Probability distribution of state at period 0 conditional on the information up to
+                                   period 0.
+
+##### Returns
+- `nothing`
+"""
+function smooth!(rsm::RegimeSwitchingModel,
+                 p_s_joint_smoothed::AbstractArray;
+                 prob_update_pre::AbstractArray = stationary_distributions(MarkovChain(rsm.P))[1],
+                 verbose::Bool=false)
+    y = rsm.y
+    M = rsm.M
+    T = size(y, 1)
+    rsm_tmp = RegimeSwitchingModel(rsm.g, y, rsm.parameter, rsm.P)
+    ps, p_s_update, p_s_forecast = filter!(rsm_tmp, prob_update_pre = prob_update_pre)
+    p_s_init = Vector{Float64}(undef, M)
+    rsm.prob_smoothed[T, :] = p_s_update[T, :]
+    for t = T-1:-1:1
+        for s in 1:M
+            for sp in 1:M
+                p_s_joint_smoothed[t, s, sp] = rsm.prob_smoothed[t+1, sp] * p_s_update[t, s] * rsm.P[s, sp]/p_s_forecast[t+1, sp]
+            end
+            rsm.prob_smoothed[t, s] = sum(p_s_joint_smoothed[t, s, :])
+        end
+    end
+    verbose ? (return p_s_joint_smoothed) : (return nothing) 

```

Comment:
Does this if statement in return cause confusion to the compiler? This could prevent it from inferring the type of the output of this function. Is there a reason to not just return `p_s_joint_smoothed`?

Reply:
I don't think it does create a problem -- It identifies it as a union of nothing and the array type:

```
julia> @code_warntype test(rsm, true)
Variables
  #self#::Core.Compiler.Const(test, false)
  rsm::Main.QuantEcon.RegimeSwitchingModel{Array{Float64,2},Array{Float64,2},Array{Float64,2}}
  verbose::Bool

Body::Union{Nothing, Array{Float64,1}}
1 ─ %1 = Main.randn(10)::Array{Float64,1}
│   %2 = (:verbose,)::Core.Compiler.Const((:verbose,), false)
│   %3 = Core.apply_type(Core.NamedTuple, %2)::Core.Compiler.Const(NamedTuple{(:verbose,),T} where T<:Tuple, false)
│   %4 = Core.tuple(verbose)::Tuple{Bool}
│   %5 = (%3)(%4)::NamedTuple{(:verbose,),Tuple{Bool}}
│   %6 = Main.QuantEcon.smooth!::Core.Compiler.Const(Main.QuantEcon.smooth!, false)
│   %7 = Core.kwfunc(%6)::Core.Compiler.Const(getfield(Main.QuantEcon, Symbol("#kw##smooth!"))(), false)
│   %8 = Main.QuantEcon.smooth!::Core.Compiler.Const(Main.QuantEcon.smooth!, false)
│   %9 = (%7)(%5, %8, rsm, %1)::Union{Nothing, Array{Float64,1}}
└──      return %9

julia> @code_warntype test(rsm, false)
Variables
  #self#::Core.Compiler.Const(test, false)
  rsm::Main.QuantEcon.RegimeSwitchingModel{Array{Float64,2},Array{Float64,2},Array{Float64,2}}
  verbose::Bool

Body::Union{Nothing, Array{Float64,1}}
1 ─ %1 = Main.randn(10)::Array{Float64,1}
│   %2 = (:verbose,)::Core.Compiler.Const((:verbose,), false)
│   %3 = Core.apply_type(Core.NamedTuple, %2)::Core.Compiler.Const(NamedTuple{(:verbose,),T} where T<:Tuple, false)
│   %4 = Core.tuple(verbose)::Tuple{Bool}
│   %5 = (%3)(%4)::NamedTuple{(:verbose,),Tuple{Bool}}
│   %6 = Main.QuantEcon.smooth!::Core.Compiler.Const(Main.QuantEcon.smooth!, false)
│   %7 = Core.kwfunc(%6)::Core.Compiler.Const(getfield(Main.QuantEcon, Symbol("#kw##smooth!"))(), false)
│   %8 = Main.QuantEcon.smooth!::Core.Compiler.Const(Main.QuantEcon.smooth!, false)
│   %9 = (%7)(%5, %8, rsm, %1)::Union{Nothing, Array{Float64,1}}
└──      return %9
```

It does add "complexity" relative to `return p_s_joint_smoothed` though, so is there a reason not to return it always?

Reply:
>  is there a reason not to return it always?

No, when I realized `p_s_joint_smoothed` should be returned, I left `return nothing` case so that I didn't need to make change to existing parts, but that was a silly idea.  Now I agree with you, it's just confusing. I'll fix it.